# OOP

We're going to look at some bank account classes.


In [16]:
class BaseAccount:
    
    def __init__(self, name, initial_deposit=0): 
        self._name = name 
        self._balance = initial_deposit 

    def name(self):
        return self._name

    def balance(self):
        return self._balance

    def withdraw(self, amount):
        self._balance -= amount
        return self._balance
    
    def deposit(self, amount):
        self._balance += amount
        return self._balance


In [17]:
my_account = BaseAccount('CS88', 100)

In [18]:
print(my_account)
my_account.withdraw(10)

my_account.balance()

90

In [15]:
my_account._balance

60

# clicker question

## What is the value of `data8.balance()` after executing the following lines?

```
cs88 = BaseAccount('CS88', 100)
data8 = BaseAccount('DATA8', 100)
cs88.deposit(25)
data8.withdraw(25)
```

### A) 100
### B) 75
### C) 125
### D) Error
### E) "I'm confused, let's review!"

# Class Attributes


In [64]:
class BaseAccount2:
    account_number_seed = 1000
    accounts_list = []
    
    def __init__(self, name, initial_deposit=0): 
        self._name = name 
        self._balance = initial_deposit 
        self._acct_no = BaseAccount2.account_number_seed     
        BaseAccount2.account_number_seed += 1
        BaseAccount2.accounts_list += [self]

    def name(self):
        return self._name

    def balance(self):
        return self._balance

    def withdraw(self, amount):
        self._balance -= amount
        return self._balance
    
    def deposit(self, amount):
        self._balance += amount
        return self._balance

    def account_no(self):
        return self._acct_no

    # Can optionally use @classmethod
    def accounts():
        return BaseAccount2.accounts_list
    
    def show_accounts():
        for account in BaseAccount2.accounts():
            print(account.name(),
                  account.account_no(),account.balance())



In [65]:
data100 = BaseAccount2('DATA100')
cs61b = BaseAccount2('CS61B')

In [66]:
print(cs61b.account_no())
print(data100.account_no())

1001
1000


In [67]:
print(BaseAccount2.show_accounts())

DATA100 1000 0
CS61B 1001 0
None


## What is the value of `BaseAccount2.account_number_seed` after executing the following lines?

```
data100 = BaseAccount2('DATA100')
cs61b = BaseAccount2('CS61B')
```

### A) 1000
### B) 1001
### C) 1002
### D) 1003
### E) "I'm confused, let's review!"

In [70]:
BaseAccount2.account_number_seed

1002